nomos-libp2p/src/behaviour/nat/state_machine.rs

```diff
@@ -0,0 +1,749 @@
+                    addr,
+                    self.state.addr_to_test(),
+                );
+                self
+            }
+            _ => self,
+        }
+    }
+}
+
+impl OnEvent<Event> for State<Public, Event> {
+    fn on_event(self: Box<Self>, event: Event, command_tx: &CommandTx) -> Box<dyn OnEvent<Event>> {
+        match event {
+            Event::ExternalAddressConfirmed(addr) | Event::AutonatClientTestOk(addr)
+                if self.state.addr() == &addr =>
+            {
+                command_tx.send(Command::ScheduleAutonatClientTest(addr));
+                self
+            }
+            Event::AutonatClientTestFailed(addr) if self.state.addr() == &addr => {
+                self.boxed(Public::into_test_if_public)
+            }
+            Event::ExternalAddressConfirmed(addr) => {
+                tracing::error!(
+                    "Swarm confirmed external address {}, but {} was expected",
+                    addr,
+                    self.state.addr(),
+                );
+                self
+            }
+            Event::AutonatClientTestOk(addr) | Event::AutonatClientTestFailed(addr) => {
+                tracing::error!(
+                    "Autonat client reported address {}, but {} was expected",
+                    addr,
+                    self.state.addr(),
+                );
+                self
+            }
+            _ => self,
+        }
+    }
+}
+
+impl OnEvent<Event> for State<MappedPublic, Event> {
+    fn on_event(self: Box<Self>, event: Event, command_tx: &CommandTx) -> Box<dyn OnEvent<Event>> {
+        match event {
+            Event::ExternalAddressConfirmed(addr) | Event::AutonatClientTestOk(addr)
+                if self.state.addr() == &addr =>
+            {
+                command_tx.send(Command::ScheduleAutonatClientTest(addr));
+                self
+            }
+            Event::AutonatClientTestFailed(addr) if self.state.addr() == &addr => {
+                self.boxed(MappedPublic::into_test_if_public)
+            }
+            Event::ExternalAddressConfirmed(addr) => {
+                tracing::error!(
+                    "Swarm confirmed external address {}, but {} was expected",
+                    addr,
+                    self.state.addr(),
+                );
+                self
+            }
+            Event::AutonatClientTestOk(addr) | Event::AutonatClientTestFailed(addr) => {
+                tracing::error!(
+                    "Autonat client reported address {}, but {} was expected",
+                    addr,
+                    self.state.addr(),
+                );
+                self
+            }
+            _ => self,
+        }
+    }
+}
+
+impl OnEvent<Event> for State<Private, Event> {
+    fn on_event(self: Box<Self>, event: Event, _: &CommandTx) -> Box<dyn OnEvent<Event>> {
+        match event {
+            Event::LocalAddressChanged(addr) => self.boxed(|state| state.into_test_if_public(addr)),
```

Comment:
Are we guaranteed the new address is different than the old one? Should we skip in the unlikely case the new matches the old?